eo-phi-normalizer/src/Language/EO/Phi/Metrics.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE OverloadedRecordDot #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{- FOURMOLU_DISABLE -}
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+
+{- FOURMOLU_ENABLE -}
+
+module Language.EO.Phi.Metrics where
+
+import Control.Lens ((+=))
+import Control.Monad (forM_)
+import Control.Monad.State (State, runState)
+import Data.Aeson (FromJSON)
+import Data.Aeson.Types (ToJSON)
+import Data.Foldable (fold)
+import Data.Function ((&))
+import Data.Generics.Labels ()
+import Data.Maybe (catMaybes)
+import Data.Traversable (forM)
+import GHC.Generics (Generic)
+import Language.EO.Phi.Rules.Common ()
+import Language.EO.Phi.Syntax.Abs
+
+data ObjectMetrics = ObjectMetrics
+  { dataless :: Int
+  , applications :: Int
+  , formations :: Int
+  , dispatches :: Int
+  }
+  deriving (Generic, Show, FromJSON, ToJSON, Eq)
+
+defaultObjectMetrics :: ObjectMetrics
+defaultObjectMetrics =
+  ObjectMetrics
+    { dataless = 0
+    , applications = 0
+    , formations = 0
+    , dispatches = 0
+    }
+
+instance Semigroup ObjectMetrics where
+  (<>) :: ObjectMetrics -> ObjectMetrics -> ObjectMetrics
+  ObjectMetrics
+    { dataless = dataless1
+    , applications = applications1
+    , formations = formations1
+    , dispatches = dispatches1
+    }
+    <> ObjectMetrics
+      { dataless = dataless2
+      , applications = applications2
+      , formations = formations2
+      , dispatches = dispatches2
+      } =
+      ObjectMetrics
+        { dataless = dataless1 + dataless2
+        , applications = applications1 + applications2
+        , formations = formations1 + formations2
+        , dispatches = dispatches1 + dispatches2
+        }
+
+instance Monoid ObjectMetrics where
+  mempty :: ObjectMetrics
+  mempty = defaultObjectMetrics
+
+data BindingMetrics = BindingMetrics
+  { name :: String
+  , metrics :: ObjectMetrics
+  }
+  deriving (Show, Generic, FromJSON, ToJSON, Eq)
+
+class Inspectable a where
+  inspect :: a -> State ObjectMetrics (Maybe Int)
+
+count :: (a -> Bool) -> [a] -> Int
+count x = length . filter x
+
+getHeight :: (Ord b, Num b) => [Binding] -> [Maybe b] -> b
+getHeight bindings heights =
+  let
+    heightAttributes = heights & catMaybes & (\case [] -> 0; x -> minimum x + 1)
+    hasDeltaAttribute = not $ null [x | x@(DeltaBinding _, _) <- zip bindings heights]
+   in
+    if hasDeltaAttribute
+      then 1
+      else heightAttributes
+
+countDataless :: Int -> Int
+countDataless x
+  | x == 0 || x > 2 = 1
+  | otherwise = 0
+
+instance Inspectable Binding where
+  inspect :: Binding -> State ObjectMetrics (Maybe Int)
```

Comment:
Do you need to repeat the type in all instances?

Reply:
It does not hurt to have the signature explicit.

Reply:
Addressed in [63f1829](https://github.com/objectionary/normalizer/pull/193/commits/63f1829e2ccf0e52fc1c8a7b23d809c049bb04a1)